echopop/utils/operations.py

```diff
@@ -239,69 +239,100 @@ def stretch(
-    # ---- On
     inner_on_lst = [inner_on] if isinstance(inner_on, str) else inner_on
     outer_on_lst = [outer_on] if isinstance(outer_on, str) else outer_on
 
-    # Merge the dataframes that will be joined to the original dataframe
-    filtered_dataframes = [df for df in df_lst if any(col in df.columns for col in inner_on_lst)]
+    # If inner_on is None, find common columns across all dataframes_to_add
+    if not inner_on:
+        common_columns_inner = set.intersection(*(set(df.columns) for df in dataframes_to_add))
+        inner_on_lst = list(common_columns_inner)
 
-    # Excluded dataframes
-    excluded_dataframes = [
-        df for df in df_lst if any(col not in df.columns for col in inner_on_lst)
-    ]
+    # Merge dataframes within dataframes_to_add on inner_on
+    if inner_on_lst:
+        merged_inner_frames = reduce(
+            lambda left, right: pd.merge(left, right, on=inner_on_lst, how=how), dataframes_to_add
+        )
+    else:
+        merged_inner_frames = pd.DataFrame()
 
-    # Begin inner merge
-    innermost_frames_to_add = reduce(
-        lambda left, right: pd.merge(left, right, on=inner_on_lst + outer_on_lst, how=how),
-        filtered_dataframes,
-    )
-    inner_frames_to_add = reduce(
-        lambda left, right: pd.merge(left, right, on=outer_on_lst, how=how),
-        excluded_dataframes + [innermost_frames_to_add],
-    )
+    # If outer_on is None, find common columns between dataframe and merged_inner_frames
+    if not outer_on:
+        common_columns_outer = set(dataframe.columns).intersection(set(merged_inner_frames.columns))
+        outer_on_lst = list(common_columns_outer)
 
-    # Find union of column names that will be used to join everything
-    union_lst = dataframe.filter(items=inner_frames_to_add.columns).columns.tolist()
+    # Merge dataframe with merged_inner_frames
+    merged_frame = dataframe.merge(merged_inner_frames, on=outer_on_lst, how=how)
 
-    # Merge together and remove NaN values depending on argument 'drop_na'
+    # Perform merge with drop_na option
     if drop_na:
-        merged_frame = dataframe.dropna().merge(
-            inner_frames_to_add.dropna(), on=union_lst, how="outer"
-        )
-    else:
-        merged_frame = dataframe.merge(inner_frames_to_add, on=union_lst, how="outer")
+        merged_frame = merged_frame.dropna()
+
+    # Restore original dtypes of columns in merged_frame
+    for col in merged_frame.columns:
+        dtype_to_restore = unique_dtypes.get(col)
+        if dtype_to_restore is not None:
+            if pd.api.types.is_integer_dtype(dtype_to_restore):
+                # Check if column contains NaN or inf values
+                if merged_frame[col].isnull().any() or not merged_frame[col].notna().all():
+                    continue  # Skip conversion if NaN or inf present
+                merged_frame[col] = merged_frame[col].astype(dtype_to_restore)
```

Comment:
This seems super specific and would probably cause some issues depending on the data it encounters. I know that NaN cannot exist as an integer, but I guess the question(s) may be:
- if the values need to be integer, why is there NaN in the first place, and can those be 0?
- does it need to be integer, can we use float (1.0, 2.0, etc) so that NaN is retained?

Reply:
This is necessary due to how `pandas.merge()` converts `int` to `float` when there are potentially mismatched records (i.e. exist in one but not another DataFrame) anywhere. So this affects the entire DataFrame rather than specific columns where there may be disagreement/conflict/contain `NaN` or `inf`. Where `NaN` may occur are other cases where there is some mismatch in the merge keys (or having a missing value to begin with) across dataframes. But `int` values needs to remain `int` to preserve various indexing functions that occur later on in the package.